core/mm/core_mmu.c

```diff
@@ -0,0 +1,2298 @@
+ * TEE_RAM_VA_SIZE should be covered, either with a direct mapping or empty
+ * mapping for pager usage. This adds translation tables as needed for the
+ * pager to operate.
+ */
+static void add_pager_vaspace(struct tee_mmap_region *mmap, size_t num_elems,
+			      size_t *last)
+{
+	paddr_t begin = 0;
+	paddr_t end = 0;
+	size_t size = 0;
+	size_t pos = 0;
+	size_t n = 0;
+
+	if (*last >= (num_elems - 1)) {
+		EMSG("Out of entries (%zu) in memory map", num_elems);
+		panic();
+	}
+
+	for (n = 0; !core_mmap_is_end_of_table(mmap + n); n++) {
+		if (map_is_tee_ram(mmap + n)) {
+			if (!begin)
+				begin = mmap[n].pa;
+			pos = n + 1;
+		}
+	}
+
+	end = mmap[pos - 1].pa + mmap[pos - 1].size;
+	size = TEE_RAM_VA_SIZE - (end - begin);
+	if (!size)
+		return;
+
+	assert(pos <= *last);
+	memmove(mmap + pos + 1, mmap + pos,
+		sizeof(struct tee_mmap_region) * (*last - pos));
+	(*last)++;
+	memset(mmap + pos, 0, sizeof(mmap[0]));
+	mmap[pos].type = MEM_AREA_PAGER_VASPACE;
+	mmap[pos].va = 0;
+	mmap[pos].size = size;
+	mmap[pos].region_size = SMALL_PAGE_SIZE;
+	mmap[pos].attr = core_mmu_type_to_attr(MEM_AREA_PAGER_VASPACE);
+}
+
+static void check_sec_nsec_mem_config(void)
+{
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(secure_only); n++) {
+		if (pbuf_intersects(nsec_shared, secure_only[n].paddr,
+				    secure_only[n].size))
+			panic("Invalid memory access config: sec/nsec");
+	}
+}
+
+static size_t collect_mem_ranges(struct tee_mmap_region *memory_map,
+				 size_t num_elems)
+{
+	const struct core_mmu_phys_mem *mem = NULL;
+	size_t last = 0;
+
+	for (mem = phys_mem_map_begin; mem < phys_mem_map_end; mem++) {
+		struct core_mmu_phys_mem m = *mem;
+
+		/* Discard null size entries */
+		if (!m.size)
+			continue;
+
+		/* Only unmapped virtual range may have a null phys addr */
+		assert(m.addr || !core_mmu_type_to_attr(m.type));
+
+		add_phys_mem(memory_map, num_elems, &m, &last);
+	}
+
+	if (IS_ENABLED(CFG_SECURE_DATA_PATH))
+		verify_special_mem_areas(memory_map, num_elems,
+					 phys_sdp_mem_begin,
+					 phys_sdp_mem_end, "SDP");
+
+	add_va_space(memory_map, num_elems, MEM_AREA_RES_VASPACE,
+		     CFG_RESERVED_VASPACE_SIZE, &last);
+
+	add_va_space(memory_map, num_elems, MEM_AREA_SHM_VASPACE,
+		     SHM_VASPACE_SIZE, &last);
+
+	memory_map[last].type = MEM_AREA_END;
+
+	return last;
+}
+
+static void assign_mem_granularity(struct tee_mmap_region *memory_map)
+{
+	struct tee_mmap_region *map = NULL;
+
+	/*
+	 * Assign region sizes, note that MEM_AREA_TEE_RAM always uses
+	 * SMALL_PAGE_SIZE.
+	 */
+	for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
+		paddr_t mask = map->pa | map->size;
+
+		if (!(mask & CORE_MMU_PGDIR_MASK))
+			map->region_size = CORE_MMU_PGDIR_SIZE;
+		else if (!(mask & SMALL_PAGE_MASK))
+			map->region_size = SMALL_PAGE_SIZE;
+		else
+			panic("Impossible memory alignment");
+
+		if (map_is_tee_ram(map))
+			map->region_size = SMALL_PAGE_SIZE;
+	}
+}
+
+static unsigned int get_va_width(void)
```

Comment:
Maybe move this to arch instead ? not sure about `IS_ENABLED(ARM64)`  since we get `error: ‘CFG_LPAE_ADDR_SPACE_BITS’ undeclared`

Reply:
OK